Nautilus/MonoBehaviours/EntitySpawner.cs

```diff
@@ -1,74 +1,106 @@
-
-        LargeWorldEntity lwe = obj.GetComponent<LargeWorldEntity>();
-
         LargeWorldStreamer lws = LargeWorldStreamer.main;
         yield return new WaitUntil(() => lws != null && lws.IsReady()); // first we make sure the world streamer is initialized
 
-        // non-global objects cannot be spawned in unloaded terrain so we need to wait
-        if (lwe is {cellLevel: not (LargeWorldEntity.CellLevel.Batch or LargeWorldEntity.CellLevel.Global)})
+        if (!global)
         {
-            Int3 batch = lws.GetContainingBatch(spawnInfo.SpawnPosition);
-            yield return new WaitUntil(() => lws.IsBatchReadyToCompile(batch)); // then we wait until the terrain is fully loaded (must be checked on each frame for faster spawns)
+            // then we wait until the terrain is fully loaded (must be checked on each frame for faster spawns)
+            yield return new WaitUntil(() => lws.IsBatchReadyToCompile(batchId));
         }
 
         LargeWorld lw = LargeWorld.main;
-
+        
         yield return new WaitUntil(() => lw != null && lw.streamer.globalRoot != null); // need to make sure global root is ready too for global spawns.
+
+        foreach (var spawnInfo in spawnInfos)
+        {
+            string stringToLog = spawnInfo.Type switch
+            {
+                SpawnInfo.SpawnType.ClassId => spawnInfo.ClassId,
+                _ => spawnInfo.TechType.AsString()
+            };
+            
+            InternalLogger.Debug($"Spawning {stringToLog}");
             
-        lw.streamer.cellManager.RegisterEntity(obj);
+            TaskResult<GameObject> task = new();
+            yield return GetPrefabAsync(spawnInfo, task);
+
+            GameObject prefab = task.Get();
+            if (prefab == null)
+            {
+                InternalLogger.Error($"no prefab found for {stringToLog}; process for Coordinated Spawn canceled.");
+                continue;
+            }
+
+            LargeWorldEntity lwe = prefab.GetComponent<LargeWorldEntity>();
+
+            if (!lwe)
+            {
+                // @Metious: I don't think this is necessary because LargeWorldEntity.Register(obj) Ensures that the object has a LargeWorldEntity component so we could replace 
+                // lwe.cellLevel < LargeWorldEntity.CellLevel.Batch with (lwe == null || lwe.cellLevel < LargeWorldEntity.CellLevel.Batch) and it should work fine.
+                InternalLogger.Error($"No LargeWorldEntity component found for prefab '{stringToLog}'; process for Coordinated Spawn canceled.");
+                continue;
+                // 😎 Nice.
+            }
+
+            if (lwe.cellLevel < LargeWorldEntity.CellLevel.Batch && !lws.IsRangeActiveAndBuilt(new Bounds(spawnInfo.SpawnPosition, Vector3.one)))
+            {
+                // Cells aren't ready yet. We have to wait until they are.
+                StartCoroutine(WaitForCellLoaded(lws, prefab, spawnInfo, stringToLog));
+                continue;
+            }
```

Comment:
@Metious can I get you to check this part out because the game itself Ensures a LWE on any registered object and just uses the default cell level which would be near I think

Reply:
It always has been the case, but I dislike it because we're assuming cell levels, which are quite important on how the object loads/saves. I think since we always had them error before, it's better to keep it that way so users still get notified for possible mistakes and also backward compatibility.

Also `LargeWorldStreamer.IsRangeActiveAndBuilt` depends on the clipmap being ready, which is extremely slow and kinda broken especially in the void. In my experience it always returned false no matter how much I waited in the spawning zone, but it only returns true when I teleport to the exact spawn position.
I was using that before switching to just checking for the cells being ready; in my tests in SN cells are almost immediately initialized after the batch is ready for compilation, so if BZ doesn't have the method, they probably don't need them anymore due to that, so maybe keep the cell check for SN only?

Reply:
Before you actually had a zero sized bounds for some reason and were using it based on the batch center.

I converted this to actually use the same code the game itself uses to check the location
By checking a vector3.one×5 at the spawn location instead of the batch Location. 

I will do some testing for void spawns with my way to see.

Reply:
I changed it to batch center only after making the entity spawner per-batch. During my tests I was using the spawn info's position and checked it against 0 size, 3 size and 5 size. All of them returned false when they should be spawning. 0 sized bounds is the most precise because it only checks for the exact spawn position's cells and nothing else.

I will test your code again and report back, but the game never waits for the clipmap to be ready to spawn stuff.

Reply:
Just tested your code and can confirm that the issue I mentioned exists.

In the void, before my object that is supposed to spawn at `373 -400 1920` was being spawned when I was at `372 -395 1801`, but with your code I had to swim to the exact spawning position.

The same is true for spawns that are in the crater, but you don't have to go inside the spawning position; just a little bit closer to the spawn position but the spawn will still surprise you have to get awfully close.